app/src/main/java/rightmesh/left/io/gpstracker/MainActivity.java

```diff
@@ -61,20 +48,56 @@ protected void onCreate(Bundle savedInstanceState) {
 
         tvNotification = findViewById(R.id.tv_notification);
 
-        Logger.log(TAG, "ON CREATE");
+        initViewModel(savedInstanceState);
+        observeViewModel();
 
-        androidMeshManager = AndroidMeshManager
-                .getInstance(MainActivity.this,
-                        MainActivity.this,
-                        null,
-                        SUPER_PEER_URL);
+        Logger.log(TAG, "ON CREATE");
 
         locationUtil = new LocationTracker(this, getLifecycle())
                 .setInterval(1000)
                 .setFastestInterval(500)
                 .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
 
+        askPermission();
+    }
+
+    /**
+     * Init viewmodel.
+     *
+     * @param savedInstanceState savedInstanceState – If the activity is being re-initialized after
+     *                           previously being shut down then this Bundle contains the data it
+     *                           most recently supplied in
+     */
+    private void initViewModel(Bundle savedInstanceState) {
+        viewModel = ViewModelProviders.of(this).get(MainViewModel.class);
+        if (savedInstanceState == null) {
+            viewModel.init();
+        }
+    }
+
+    /**
+     * Binding data from viewmodel to UI.
+     */
+    private void observeViewModel() {
+        viewModel.liveDataNotificationText.observe(this, s -> {
```

Comment:
Minor: what is 's'?

Reply:
it's a text to show in Textview. I've just changed it to notificationText to be more readable